clickhouse_connect/driverc/buffer.pyx

```diff
@@ -0,0 +1,223 @@
+        self.end = 0
+        chunk = next(self.gen)
+        x = len(chunk)
+        if x == 0:
+            raise IndexError
+        py_chunk = chunk
+        if x > 1:
+            self._reset_buff(chunk)
+            self.buf_loc = 1
+            self.end = x
+        return <unsigned char>chunk[0]
+
+    @cython.inline
+    cdef char* _read_bytes(self, unsigned long long sz) except NULL:
+        if self._set_slice(sz) == 255:
+            return NULL
+        return self._cur_slice()
+
+    @cython.inline
+    cdef object _read_leb128_str(self, char* encoding='utf8'):
+        cdef unsigned long long sz = self.read_leb128()
+        cdef char* b = self._read_bytes(sz)
+        if b == NULL:
+            raise StopIteration
+        try:
+            return PyUnicode_Decode(b, sz, encoding, errors)
+        except UnicodeDecodeError:
+            return PyBytes_FromStringAndSize(b, sz).hex()
+
+    def read_leb128_str(self, encoding: str='utf8'):
+        return self._read_leb128_str(encoding.encode())
+
+    def read_byte(self) -> int:
+        cdef unsigned char ret = self._read_byte()
+        if ret == 255 and PyErr_Occurred():
+            raise StopIteration
+        return ret
+
+    @cython.inline
+    cpdef unsigned long long read_leb128(self) except? 99999:
+        cdef:
+            unsigned long long sz = 0, shift = 0
+            unsigned char b
+        while 1:
+            b = self._read_byte()
+            if b == 255 and PyErr_Occurred():
+                raise StopIteration
+            sz += ((b & 0x7f) << shift)
+            if (b & 0x80) == 0:
+                return sz
+            shift += 7
+
+    def read_uint64(self) -> int:
+        cdef ull_wrapper* x
+        if self._set_slice(8) == 255:
+            raise StopIteration
+        if must_swap:
+            memcpy(swapper.data.as_voidptr, self._cur_slice(), 8)
+            swapper.byteswap()
+            return swapper[0]
+        x = <ull_wrapper *> self._cur_slice()
+        return x.int_value
+
+    def read_bytes(self, unsigned long long sz):
+        if self._set_slice(sz) == 255:
+            raise StopIteration
+        return self._cur_slice()[:sz]
+
+    def read_str_col(self, unsigned long long num_rows, encoding: str = 'utf8'):
+        cdef object column = PyTuple_New(num_rows)
+        cdef char* enc
+        pyenc = encoding.encode()
+        enc = pyenc
+        for x in range(num_rows):
+            v = self._read_leb128_str(enc)
+            PyTuple_SET_ITEM(column, x, v)
+            Py_INCREF(v)
+        return column
+
+    def read_array(self, t: str, unsigned long long num_rows):
+        cdef array.array template = array_templates[t]
+        cdef array.array result = array.clone(template, num_rows, 0)
+        cdef unsigned long long sz = result.itemsize * num_rows
+        if self._set_slice(sz) == 255:
+            raise StopIteration
+        memcpy(result.data.as_voidptr, self._cur_slice(), sz)
+        if must_swap:
```

Comment:
does it work without `global must_swap`? If Cython is alike normal Python from what I recall, this might be `false` all the time